internal/character/bailu/eidolon.go

```diff
@@ -0,0 +1,60 @@
+package bailu
+
+import (
+	"github.com/simimpact/srsim/pkg/engine/event"
+	"github.com/simimpact/srsim/pkg/engine/info"
+	"github.com/simimpact/srsim/pkg/engine/modifier"
+	"github.com/simimpact/srsim/pkg/engine/prop"
+	"github.com/simimpact/srsim/pkg/key"
+	"github.com/simimpact/srsim/pkg/model"
+)
+
+const (
+	E4 key.Modifier = "bailu-e4"
+)
+
+func init() {
+	modifier.Register(E4, modifier.Config{
+		Stacking:          modifier.Replace,
+		MaxCount:          3,
+		CountAddWhenStack: 1,
+		StatusType:        model.StatusType_STATUS_BUFF,
+		CanModifySnapshot: true,
+	})
+}
+
+func (c *char) initEidolons() {
+	// E1 : If the target ally's current HP is equal to their Max HP when
+	// 			Invigoration ends, regenerates 8 extra Energy for this target.
+	c.engine.Events().ModifierRemoved.Subscribe(func(e event.ModifierRemoved) {
+		// negative checks
+		if e.Modifier.Name != invigoration ||
+			c.info.Eidolon < 1 ||
+			c.engine.HPRatio(e.Target) != 1.0 {
+			return
+		}
+		// add flat energy
+		c.engine.ModifyEnergy(info.ModifyAttribute{
+			Key:    invigoration,
+			Target: e.Target,
+			Source: c.id,
+			Amount: 8.0,
+		})
+	})
+
+	// E4 : Every healing provided by the Skill makes the recipient deal
+	// 10% more DMG for 2 turn(s). This effect can stack up to 3 time(s).
+	c.engine.Events().HealEnd.Subscribe(func(e event.HealEnd) {
+		if c.info.Eidolon < 4 {
+			return
+		}
+		if e.Healer == c.id && e.Key == Skill {
+			c.engine.AddModifier(e.Target, info.Modifier{
+				Name:     E4,
+				Source:   c.id,
+				Stats:    info.PropMap{prop.AllDamagePercent: 0.1},
+				Duration: 2,
+			})
+		}
```

Comment:
I believe instead of a `HealEnd` event subscription, you can instead use a listener for `OnAfterDealHeal` similar to the DM.  This is because the set of actions we care about in this case is limited to strictly actions dealt by Bailu.

Reply:
about using event subscription instead of a modifier with a listener, I do that because :
1. Tbh I tried using a listener and it ended up being longer (more lines of codes). I thought as long as I do the proper check (i.e. checking for caster is bailu, etc) it should be equivalent to a listener version.
2. Iirc my Seele PR also had some of these and it's still approved so I thought it's fine.
3. It increases the amount of modifiers. It's hardly an issue tho, just adds more stuff in the logs.
If you still want me to refactor it then I'll do it though dw.
I'll merge the early return conditions into 1 if statement while I'm at it too.